daemon/emer-aggregate-tally.c

```diff
@@ -315,101 +340,74 @@ emer_aggregate_tally_iter (EmerAggregateTally *self,
-                   tallies_path, error->message);
-      return;
-    }
+  CHECK (sqlite3_prepare_v2 (self->db, SELECT_SQL, -1, &stmt, NULL));
+  CHECK (sqlite3_bind_text (stmt, 1, date, -1, SQLITE_TRANSIENT));
 
-  while (TRUE)
+  while ((ret = sqlite3_step (stmt)) == SQLITE_ROW)
     {
-      g_autoptr(GVariant) variant = NULL;
-      g_autoptr(GVariant) event_id = NULL;
-      g_autoptr(GVariant) aggregate_key = NULL;
-      g_autoptr(GVariant) payload = NULL;
-      g_autoptr(GError) local_error = NULL;
-      g_autofree gchar *path = NULL;
+      g_autoptr(GVariant) event_id =
+        text_to_variant (stmt, 1);
+      guint32 unix_user_id = column_to_uint32 (stmt, 2);
+      const gchar *aggregate_key_type =
+        (const gchar *) sqlite3_column_text (stmt, 3);
+      g_autoptr(GVariant) aggregate_key =
+        column_to_variant (stmt, aggregate_key_type, 4);
+      const gchar *payload_type =
+        (const gchar *)sqlite3_column_text (stmt, 5);
+      g_autoptr(GVariant) payload =
+        column_to_variant (stmt, payload_type, 6);
+      guint32 counter = column_to_uint32 (stmt, 7);
       EmerTallyIterResult result;
-      GFileInfo *info;
-      GFile *file;
-      guint32 counter = 0;
-      guint32 unix_user_id = 0;
-
-      g_file_enumerator_iterate (enumerator, &info, &file, NULL, &local_error);
-
-      if (local_error)
-        {
-          if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
-            g_warning ("Error listing tallies from %s: %s",
-                       tallies_path, local_error->message);
-          return;
-        }
-
-      if (!file || !info)
-        break;
-
-      if (g_file_info_get_file_type (info) != G_FILE_TYPE_REGULAR)
-        continue;
-
-      path = g_file_get_path (file);
-
-      variant = load_aggregate_timer_from_path (path);
-      if (!variant)
-        continue;
-
-      g_variant_get (variant,
-                     AGGREGATE_TIMER_VARIANT_TYPE,
-                     &event_id,
-                     &aggregate_key,
-                     &unix_user_id,
-                     &counter,
-                     &payload);
 
       result = func (unix_user_id, event_id,
                      aggregate_key, payload,
                      counter, date, user_data);
 
       if (flags & EMER_TALLY_ITER_FLAG_DELETE)
         {
-          g_autoptr(GError) deletion_error = NULL;
-
-          if (!g_file_delete (file, NULL, &deletion_error))
-            g_warning ("Error deleting tally file %s: %s",
-                       path, deletion_error->message);
+          const gint row_id = sqlite3_column_int (stmt, 0);
+          g_array_append_val (rows_to_delete, row_id);
         }
 
       if (result & EMER_TALLY_ITER_STOP)
         break;
     }
 
-  if (flags & EMER_TALLY_ITER_FLAG_DELETE)
+  CHECK (sqlite3_finalize (stmt));
+
+  if (rows_to_delete->len > 0)
     {
-      g_autoptr(GError) deletion_error = NULL;
+      const char *DELETE_SQL = "DELETE FROM tally WHERE id = ?;";
```

Comment:
An optimization that I have seen done is to generate a query of the form:

```
DELETE FROM tally WHERE id = (?, ?, ?, ...);
```

i.e. the number of `?` is dynamically generated. Then you bind all the IDs and execute a single statement. But this is ugly to implement if the sql binding doesn't provide it, (I don't think sqlite does,) and in this case I think doing it one at a time is fine.